### PR TITLE
[BUGFIX] Fix virtual and normal collections

### LINKED ISSUE #641: Combination of a virtual collection with the parameter "Nur Kollektion" in the search plugin brings a null result
## Description

We made a virtual collection "Briefe von Wolfgang Borchert" with the elements 
Aus Nachlass Wolfgang Borchert
HANSb*
for the solrquery.
A query from the serach result list (see reproduction) brings a null result for correct search terms. The solrLogging (see "Screenshots and Examples) gives information about the missing switch from the name of the virtual collection to a proper solrquery as reason for the null result.

## Reproduction

Steps to reproduce the behaviour:

1. Go to https://digitalisate.sub.uni-hamburg.de/index.php?id=2366
2. Choose "Briefe von Wolfgang Borchert"
3. Search for "aline" (Part of item 2ff in the browsing list of "Briefe von Wolfgang Borchert")
4. See error with a null result

## Expected Behavior
We see the benefit of virtual collections and want to place it with correct search results, when a restriction to collection is preset in the search plugin.

## Screenshots and Examples
**SolrLogging**
#Suche nach **aline** in der virtuellen Kollektion
2021-06-04 10:59:03.538 INFO  (qtp1979787155-89) [    x:dlfCore0] o.a.s.c.S.Request [dlfCore0]  webapp=/solr path=/admin/ping params={json.nl=flat&omitHeader=true&wt=json} hits=2358799 status=0 QTime=0
2021-06-04 10:59:03.538 INFO  (qtp1979787155-89) [    x:dlfCore0] o.a.s.c.S.Request [dlfCore0]  webapp=/solr path=/admin/ping params={json.nl=flat&omitHeader=true&wt=json} status=0 QTime=0
2021-06-04 10:59:03.631 INFO  (qtp1979787155-106) [    x:dlfCore0] o.a.s.c.S.Request [dlfCore0]  webapp=/solr path=/select params={q=aline&json.nl=flat&omitHeader=true&fl=*,score&start=0&sort=score+desc&fq=collection_faceting:("Briefe+Wolfgang+Borchert")&rows=0&wt=json} hits=0 status=0 QTime=0
2021-06-04 10:59:03.707 INFO  (qtp1979787155-102) [    x:dlfCore0] o.a.s.c.S.Request [dlfCore0]  webapp=/solr path=/select params={q={!join+from%3Duid+to%3Duid}aline&json.nl=flat&omitHeader=true&fl=uid,id&start=0&sort=score+desc&fq={!join+from%3Duid+to%3Duid}**collection_faceting:("**Briefe+Wolfgang+Borchert**")**&fq=toplevel:true&rows=0&wt=json} hits=0 status=0 QTime=0
2021-06-04 10:59:03.888 INFO  (qtp1979787155-125) [    x:dlfCore0] o.a.s.c.S.Request [dlfCore0]  webapp=/solr path=/admin/ping params={json.nl=flat&omitHeader=true&wt=json} hits=2358799 status=0 QTime=0
2021-06-04 10:59:03.888 INFO  (qtp1979787155-125) [    x:dlfCore0] o.a.s.c.S.Request [dlfCore0]  webapp=/solr path=/admin/ping params={json.nl=flat&omitHeader=true&wt=json} status=0 QTime=0
2021-06-04 10:59:03.970 INFO  (qtp1979787155-85) [    x:dlfCore0] o.a.s.c.S.Request [dlfCore0]  webapp=/solr path=/select params={f.year_faceting.facet.sort=count&facet.field={!key%3Dtype_faceting}type_faceting&facet.field={!key%3Dplace_faceting}place_faceting&facet.field={!key%3Dyear_faceting}year_faceting&facet.field={!key%3DDocLanguage_faceting}DocLanguage_faceting&facet.field={!key%3Dcollection_faceting}collection_faceting&json.nl=flat&fl=*,score&f.year_faceting.facet.limit=30&f.DocLanguage_faceting.facet.limit=30&start=0&sort=score+desc&fq=collection_faceting:("**Briefe+Wolfgang+Borchert**")&rows=0&q=aline&f.place_faceting.facet.limit=30&f.type_faceting.facet.sort=count&omitHeader=true&f.collection_faceting.facet.sort=count&f.type_faceting.facet.limit=30&f.place_faceting.facet.sort=count&f.DocLanguage_faceting.facet.sort=count&f.collection_faceting.facet.limit=30&wt=json&facet=true} hits=0 status=0 QTime=2
2021-06-04 10:59:04.060 INFO  (qtp1979787155-28) [    x:dlfCore0] o.a.s.c.S.Request [dlfCore0]  webapp=/solr path=/select params={q=aline&json.nl=flat&omitHeader=true&fl=*,score&start=0&sort=score+desc&fq=collection_faceting:("Briefe+Wolfgang+Borchert")&rows=0&wt=json} hits=0 status=0 QTime=0
2021-06-04 10:59:04.172 INFO  (qtp1979787155-90) [    x:dlfCore0] o.a.s.c.S.Request [dlfCore0]  webapp=/solr path=/select params={q={!join+from%3Duid+to%3Duid}aline&json.nl=flat&omitHeader=true&fl=uid,id&start=0&sort=score+desc&fq={!join+from%3Duid+to%3Duid}collection_faceting:("**Briefe+Wolfgang+Borchert**")&fq=toplevel:true&rows=25&wt=json} hits=0 status=0 QTime=0

## Environment

 - OS version: Debian Linux 4.19 (Host Europe/effective webwork)
 - RDBMS version:  MySQL 5.5.5
 - Apache Solr version: 7.4.0
 - TYPO3 version: 8.7.25
 - PHP version: 7.1.33

### LINKED ISSUE #866: Virtual collection does not get displayed in listview-plugin
## Description

It's possible to create user defined collections with a specific solr-query. These virtual collections get displayed within the collections-plugin and the number of documents gets properly counted via the specified solr-query. When entering the collection the listview plugin shows an empty resultset.

## Reproduction

Create a new collection, e.g.:
label = Virtual Collection
index_name = virtual_collection
index_search = type:monograph

You get the tile, but you wont get a resultset

## Reason

The solrSearch.php does not distinguish wether a collection is purely virtual and defined by a solr-query or is in fact indexed in the "collection:" field within the index. To resolve this issue, we must at least verify, if a collection has a index_search setting, otherwise we should use the collection_facetting query as intended:

\Classes\Common\SolrSearch.php https://github.com/kitodo/kitodo-presentation/blob/48403356f41efdf3e133901750b45bf783e0d33d/Classes/Common/SolrSearch.php#L233

`
<pre>
            if ($this->collection instanceof Collection) {
                // mk 2022-11-17 # virtual collection fix?
                $virtualCollectionsQueryString = $this->collection->getIndexSearch();
                $collectionsQueryString = '"' . $this->collection->getIndexName() . '"';
            } else {
</pre>

<pre>
            // mk 2022-11-17 # test virtual collection
            if(!empty($virtualCollectionsQueryString)) {
                $params['filterquery'][]['query'] = '(' . $virtualCollectionsQueryString . ')';
            } else {
                $params['filterquery'][]['query'] = 'collection_faceting:(' . $collectionsQueryString . ')';
            }
</pre>
`

## Additional Context

I cant tell, if we need to look further for collection-sets, that somehow happen it the else clause, because i havent seen such meta-collections yet. So i cant tell, if there are other side effects - this fix only works for single collections.

### PR DESCRIPTION
**Pull request by michaelkubina**
This PR fixes a variety of issues with collections:
- the collection parameter (context) does not get lost, when using facets or the sorting form
- collections are now unified in the collection parameter instead of two separate ones (= single collection and list of collections)
- list of collections in the search-plugin (settings.collections) are working now
- the facet count now works properly when entering a collection or browsing a list of collections
- virtual collections are working now and the solr-queries get applied on the actual search
- combinations of normal and virtual collections can be intertwined and will be queried and displayed correctly

### CODE REVIEW

**Comment by michaelkubina** [comment]
Fix for virtual collection issues https://github.com/kitodo/kitodo-presentation/issues/866 and https://github.com/kitodo/kitodo-presentation/issues/641

For the latter its the missing implementation as well.

File: Classes/Common/SolrSearch.php [hunk 1]
```diff
@@ -237,22 +237,36 @@ public function prepare()
...
-                foreach ($this->collection as $index => $collectionEntry) {
-                    $collectionsQueryString .= ($index > 0 ? ' OR ' : '') . '"' . $collectionEntry->getIndexName() . '"';
+            $collectionsQueryString = '';
+            $virtualCollectionsQueryString = '';
+            foreach ($this->collection as $collectionEntry) {
+                // check for virtual collections query string
+                if($collectionEntry->getIndexSearch()) {
+                    $virtualCollectionsQueryString .= empty($virtualCollectionsQueryString) ? '(' . $collectionEntry->getIndexSearch() . ')' : ' OR ('. $collectionEntry->getIndexSearch() . ')' ;
+                }
+                else {
```
Comment (sebastian-meyer) [inline Classes/Common/SolrSearch.php]:
```suggestion
                } else {
```

File: Classes/Common/SolrSearch.php [hunk 2]
```diff
@@ -237,22 +237,36 @@ public function prepare()
...
-                $params['filterquery'][]['query'] = 'toplevel:true';
-                $params['filterquery'][]['query'] = 'partof:0';
+            // virtual collections might query documents that are neither toplevel:true nor partof:0 and need to be searched separatly
+            if(!empty($virtualCollectionsQueryString)) {
+                $virtualCollectionsQueryString = '(' . $virtualCollectionsQueryString . ')';
             }
-            $params['filterquery'][]['query'] = 'collection_faceting:(' . $collectionsQueryString . ')';
+
+            // combine both querystrings into a single filterquery via OR if both are given, otherwise pass either of those
+            $params['filterquery'][]['query'] = implode(" OR ", array_filter([$collectionsQueryString, $virtualCollectionsQueryString]));
```
Comment (sebastian-meyer) [inline Classes/Common/SolrSearch.php]:
```suggestion
            $params['filterquery'][]['query'] = implode(' OR ', array_filter([$collectionsQueryString, $virtualCollectionsQueryString]));
```

**Review by sebastian-meyer — CHANGES_REQUESTED** [review]

**Review by sebastian-meyer — CHANGES_REQUESTED** [review]
This looks fine apart from the tests not running through. Please have a look at this!

**Comment by michaelkubina** [comment]
I can see the issue, but i am not sure wether i'm allowed to fix it in the testrun itself or if this has to stay unchanged and i must apply a fix to the `findSolrByCollection()` function.

Explanation: The expected type for the collection argument in `findSolrByCollection()` has changed...before it could have been either a string or an array. `solrSearch.php` has then decided how to build the query, depending on if its just a single collection or a list of collections (the latter did not work iirc). This has been now harmonised so that an array is always the expected argument.

The reason behind it is, that the `search-parameter` does not differentiate between a single collection (when coming from the collection-controller and others) or a list of collections (only when coming from the search-plugin/listview) anymore. Both use the same collection-parameter now (instead of the 2 seperate ones before) and either provides a single UID or a comma separated list of UIDs. This gets exploded, so we can loop through the array, and we obtain the collections from those UIDs and save them as a new array. The query building mechanism in the `SolrSearch.php`, that comes next, now does not need to check wether it got a string or array anymore, but can simply loop through it...which helps readabilty and made the integration of virtual collections queries easier.

Internally all calls to `findSolrByCollection` now (since this PR) pass collections as arrays, except for the testrun which i did not know of.

So, without being able to test it, i think just a small change in the `canSearchInCollection()` function is required - at least i hope so. Like:

```
$musik = array($collections[0]);
$dresdnerHefte = array($collections[1]);
```

Some feedback would be appreciated, because i dont want to mess with the unit test itself.

**Comment by sebastian-meyer** [comment]
> I can see the issue, but i am not sure wether i'm allowed to fix it in the testrun itself or if this has to stay unchanged and i must apply a fix to the `findSolrByCollection()` function.

Since you changed how collections are handled, you need to update the test accordingly.

**Review by sebastian-meyer — APPROVED** [review]
Thank you very much!

**Pull request merged**